components/hab/src/command/plan/render.rs

```diff
@@ -0,0 +1,176 @@
+
+    // import default.toml values, convert to JSON
+    // ui.begin(format!("Importing user.toml: {}", &user_toml_path));
+    let user_toml = match user_toml_path {
+        Some(path) => {
+            if !(quiet) {
+              // print helper message, maybe only print if '--verbose'? how?
+              ui.begin(format!("Importing user.toml: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+                .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        None => "".to_string(),
+    };
+    // merge default into data struct
+    merge(&mut data, toml_to_json(&user_toml));
+
+    // read mock data if provided
+    let mock_data = match mock_data_path {
+        Some(path) => {
+            if !(quiet) {
+                // print helper message, maybe only print if '--verbose'? how?
+                ui.begin(format!("Importing override file: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+              .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        // return an empty json block if '--mock-data' isn't defined.
+        // this allows us to merge an empty JSON block
+        None => "{}".to_string(),
+    };
+    // merge mock data into data
+    merge(&mut data, serde_json::from_str(&mock_data).unwrap());
+
+    // create a template renderer
+    let mut renderer = TemplateRenderer::new();
+    // register our template 
+    renderer
+        .register_template_string(&template_path, &template)
+        .expect("Could not register template content");
+    // render our JSON override in our template.
+    let rendered_template = renderer.render(&template_path, &data).ok().unwrap();
+    
+    if print {
+        if !(quiet) {
+          ui.br()?;
+          ui.warn(format!("###======== Rendered template: {}", &template_path))?;
+        }
+
+        println!("{}", rendered_template);
+
+        if !(quiet) {
+          ui.warn(format!("========### End rendered template: {}", &template_path))?;
+        }
+    }
+
+    // if not no render dir (aka "unless no_render_dir == true")
+    if !(no_render_dir) {
+      // Render our template file
+      create_with_template(ui, &format!("{}/{}", render_dir, file_name), &rendered_template, quiet)?;
+    }
+
+    if !(quiet) {
+      ui.br()?;
+    }
+    // not really sure this is correct...
+    Ok(())
+}
+
+fn toml_to_json(cfg: &str) -> Json {
+    // parse TOML string to Value
+    let toml_value = cfg.parse::<Value>().expect("Error parsing TOML");
+    // convert toml to json string
+    let toml_string = serde_json::to_string(&toml_value).expect("Error encoding JSON");
+    // convert to Json::Value
+    serde_json::from_str(&format!(r#"{{ "cfg": {} }}"#, &toml_string)).unwrap()
+}
+
+// merge two Json structs
+fn merge(a: &mut Json, b: Json) {
+    match (a, b) {
+        // not sure I understand this... 
```

Comment:
This code is definitely requires some subtle understanding of the way rust works. I'll try to explain why it needs to be as it is:

```rust
fn merge(a: &mut Json, b: Json) {
```
The first thing to note is the types of the arguments. Since `a` is a mutable reference, we can change its contents, but we can't pull subfields out into owned values, only references. On the other hand, `b` is an owned value, so we can do whatever we wish, knowing that the value is consumed by this function, preventing the caller from referencing it again.

```rust
    match (a, b) {
```
We match on the pair of the arguments because we want to do something special only if *both* of them are the `Object` variant of the `Json` (actually [`serde_json::value::Value`](https://docs.serde.rs/serde_json/value/enum.Value.html) which you imported `as Json`) type.
```rust
        (a @ &mut Json::Object(_), Json::Object(b)) => {
```
This is the case where both a and b are `Json::Object`s. We bind the two a little differently because rust cannot bind by-move and by-ref in the same pattern (see `rustc --explain E0009` for more). What we'd like to do would be:
```rust
        (Json::Object(a), Json::Object(b)) => {
```
But because `a` is a reference and `b` is a owned type, that doesn't work. We could instead take a reference to `b` and match on that:
```rust
    match (a, &b) {
```
But then we run into a different problem:
```
error[E0277]: the trait bound `std::string::String: std::convert::From<&std::string::String>` is not satisfied
   --> components/hab/src/command/plan/render.rs:192:25
    |
192 |                 merge(a.entry(k).or_insert(Json::Null), v);
    |                         ^^^^^ the trait `std::convert::From<&std::string::String>` is not implemented for `std::string::String`
```
What this is saying is that [`entry`](https://docs.serde.rs/serde_json/map/struct.Map.html#method.entry) expected `k` to be an owned `String`, but we're giving it a `&String`. This is because when we changed `b` to `&b`, the types involved in  `for (k, v) in b` changed from `(String, Value)` to `(&String, &Value)`. See the difference between [the implementation of `IntoIter`](https://docs.serde.rs/serde_json/map/struct.Map.html#impl-IntoIterator) for an owned `Map` and a `&Map`.

All this is to say, we need to deal with the fact that we wanted a reference to the `Map` inside `a`, but the owned value of the `Map` inside `b`. That's why we're binding this way:
```rust
        (a @ &mut Json::Object(_), Json::Object(b)) => {
```
Both `a` and `b` are bound "by-move"; it's confusing because the type that was "moved" into `a` happens to still be a reference. Because we couldn't destructure `a`'s inner `Map` in the match binding, that's why we need:
```rust
            let a = a.as_object_mut().unwrap();
```
The `unwrap` is necessary because the type of `a` is `&mut Json`: `as_object_mut` doesn't know it can't fail.

At this point, we have the values in `a` and `b` that we want if rust allowed us to bind the way we wanted:
```rust
        (Json::Object(a), Json::Object(b)) => {
```
it just took a bit more work.

The last bit is the other match arm: what happens if either `a` or `b` are variants *other* than `Object`:
```rust
        (a, b) => *a = b,
```

In this case, we aren't really *merging* `a` and `b`, we're just overwriting `a`'s value with `b`. We only need to dereference `a` because it's a reference and `b` isn't.

I spent a bit of time looking at an alternative approach where we could make more use of the compiler checking the types for us, and the best I could come up with is binding `a` and `b` in separate expressions. The downside is that the cases end up nested, so you either need to repeat `*a = b` several times, or have an early return:
```rust
fn merge(a: &mut Json, b: Json) {
    if let Json::Object(a_map) = a {
        if let Json::Object(b_map) = b {
            for (k, v) in b_map {
                merge(a_map.entry(k).or_insert(Json::Null), v);
            }
            return;
        }
    }
    *a = b;
}
```
I like this approach because it avoids the need for any `unwraps`, but it ultimately comes down to style which you prefer.